Setup codeception tests

yariksheptykin opened the pull request:
This PR adds codeception configuration and adds a couple of sample tests to get TDD in pimcore projects going. The goal of this work is to set the bar for writing tests in a pimcore project as low a possible. The hope is: we start seeing more pimcore-based projects with decent test coverage.

yariksheptykin committed:
Setup codeception tests

jdreesen commented on composer.json:
```diff
@@ -11,7 +11,10 @@
   },
   "require-dev": {
-    "codeception/codeception": "^4.1.12"
+    "cache/integration-tests": "^0.17.0",
```
Do we need this package for codeception tests somehow? Because it was removed recently: https://github.com/pimcore/skeleton/pull/90

jdreesen commented on composer.json:
```diff
@@ -11,7 +11,10 @@
-    "codeception/codeception": "^4.1.12"
+    "cache/integration-tests": "^0.17.0",
+    "codeception/codeception": "^4.1.12",
+    "codeception/module-symfony": "^1",
```
Pimcore doesn't use the symfony module anymore, see: pimcore/pimcore#11756 and #91

jdreesen commented on composer.json:
```diff
@@ -11,7 +11,10 @@
+    "cache/integration-tests": "^0.17.0",
+    "codeception/codeception": "^4.1.12",
+    "codeception/module-symfony": "^1",
+    "phpspec/prophecy-phpunit": "^2.0"
```
This seems to be unused in the PR, and I'm not sure that we should add it just in case the user _might_ use it. There are other mocking libraries, and not to forget Codeception's/PHPUnit's own.

yariksheptykin replied:
Good point, I don't care for the actual framework, but I strongly suggest we chose one and include an example for how to mock/stub things.

jdreesen replied:
I think I'd go with the native Codeception mocks then, as they're already there. Unless there's a reason to install an extra third party package for this.

jdreesen commented on tests/bootstrap.php:
```diff
@@ -0,0 +1,15 @@
+
+// set the used pimcore/symfony environment
+putenv('APP_ENV=test');
+$_ENV['APP_ENV'] = 'test';
```
I think we should set the env in `$_SERVER`, too.

```suggestion
$_ENV['APP_ENV'] = $_SERVER['APP_ENV'] = 'test';
```

jdreesen commented on tests/functional/bootstrap.php:
```diff
@@ -0,0 +1,8 @@
+<?php
+
+// We want to use the global bootstrap, but customize the setup a bit for running a wider-scope tests.
```
```suggestion
// We want to use the global bootstrap, but customize the setup a bit to run tests with a larger scope.
```

jdreesen commented on tests/functional/bootstrap.php:
```diff
@@ -0,0 +1,8 @@
+// We want to use the global bootstrap, but customize the setup a bit for running a wider-scope tests.
+require_once dirname(__DIR__) . '/bootstrap.php';
+
+// Let tests believe that the process is executed as a console application.
+// This helps process manager bundle from elements handle shutdown of the testsuite situation properly.
+const PIMCORE_CONSOLE = true;
```
Is this really true for *all* functional tests? What about functional controller/web tests?

yariksheptykin replied:
I am not sure about this one. It is definitely true for tests executing / simulating CLI commands. On the other hand the whole test is executed as a CLI command. I'll move this declaration into the sample command test for now.

jdreesen commented on tests/functional/Importer/CliCommandTest.php:
```diff
@@ -0,0 +1,37 @@
+
+class CliCommandTest extends KernelTestCase
+{
+    private ObjectManager $manager;
```
This is unused.

yariksheptykin commented:
Thanks for the review @jdreesen ! I'll be back :)

yariksheptykin committed:
Explain clearly

Co-authored-by: Jacob Dreesen <jacob@hdreesen.de>

yariksheptykin committed:
Declare test env in $_SERVER too

Co-authored-by: Jacob Dreesen <jacob@hdreesen.de>

yariksheptykin committed:
Remove unused stuff

yariksheptykin committed:
Add a basic unit test with a mock

yariksheptykin committed:
Declare PIMCORE_CONSOLE in CLI test

yariksheptykin committed:
Pimcore ❤ TDD

jdreesen commented on tests/unit/Controller/DefaultControllerTest.php:
```diff
@@ -0,0 +1,59 @@
+    /** @var MockObject|Environment  */
+    private MockObject $twig;
+
+    protected function setUp(): void
+    {
+        parent::setUp();
+
+        // Create a test double for twig engine.
+        // See: https://phpunit.readthedocs.io/en/9.5/test-doubles.html
+        $this->twig = $this->createMock(Environment::class);
+
+        $container = new Container();
+        $container->set('twig', $this->twig);
+        $container->set('pimcore.templating', new TwigDefaultDelegatingEngine($this->twig, []));
+
+        $this->controller = new DefaultController();
+        $this->controller->setContainer($container);
+    }
+
+    public function testDefaultAction()
+    {
+        $this->twig->method('render')->will(
+            $this->returnValueMap([
+                // Simulate rendering of default template.
+                ['default/default.html.twig', [], 'At pimcore we love writing tests! ❤️TDD!']
+            ])
+        );
+
+        $response = $this->controller->defaultAction($this->createMock(Request::class));
+
+        self::assertEquals(200, $response->getStatusCode());
+        self::assertStringContainsStringIgnoringCase('pimcore', $response->getContent());
+        self::assertStringContainsStringIgnoringCase('❤', $response->getContent());
+        self::assertStringContainsStringIgnoringCase('tests', $response->getContent());
+        self::assertStringNotContainsStringIgnoringCase('bugs', $response->getContent());
+        self::assertStringNotContainsStringIgnoringCase('hacks', $response->getContent());
+        self::assertStringNotContainsStringIgnoringCase('💩', $response->getContent());
+    }
```
I am not sure if it makes sense to test a controller in this way.... actually, [Symfony's `WebTestCase`s](https://symfony.com/doc/current/testing.html#application-tests) should be used for that. Does Codeception have an equivalent?

jdreesen replied:
But then we've nothing left to test as an example of a unit test, and that's what you wanna show, isn't it?

yariksheptykin replied:
Exactly, this test is a demonstration of how to unit-test a controller or any other class actually. WebTestCase is definitely more powerful, and we should add a demo of how to do it too. I suggest we add another suite for this kind of tests. But let's not put too much in this PR. My goal was to kick start writing tests.

aryaantony92 commented:
Could you please rebase the PR to get the latest changes and adapt the changes with codeception 5 ?

aryaantony92 commented:
> Could you please rebase the PR to get the latest changes and adapt the changes with codeception 5 ?

@yariksheptykin Just a gentle reminder

yariksheptykin commented:
Hey, @aryaantony92 ! Thanks for the reminder, I will update the PR by the end of the week.

aryaantony92 commented:
> Hey, @aryaantony92 ! Thanks for the reminder, I will update the PR by the end of the week.

Thanks 👍

jdreesen commented on composer.json:
```diff
@@ -33,11 +41,20 @@
     ],
     "post-update-cmd": [
       "Pimcore\\Composer::postUpdate",
+      "@pimcore-scripts",
+      "Pimcore\\Composer::executeMigrationsUp",
```
This has been removed here: #113

yariksheptykin replied:
Yeah, the line sneaked back in during rebase. Good catch

jdreesen commented on tests/unit/ReadmeTest.php:
```diff
@@ -0,0 +1,31 @@
+use Codeception\Test\Unit;
+
+/**
+ * This is a tongue-in-chick test created to validate codeception functionality.
```
I think you mean:

```suggestion
 * This is a tongue-in-cheek test created to validate codeception functionality.
```

yariksheptykin committed:
Chick to cheek

Co-authored-by: Jacob Dreesen <jacob@hdreesen.de>

yariksheptykin committed:
it's simply php. nice!

yariksheptykin committed:
Erase/rewind

yariksheptykin committed:
Fix tests. cs-fix.

dvesh3 commented on tests/bootstrap.php:
```diff
@@ -0,0 +1,15 @@
+
+// set the used pimcore/symfony environment
+putenv('APP_ENV=test');
+$_ENV['APP_ENV'] = $_SERVER['APP_ENV'] = 'test';
```
Isn't above line is enough for setting app env?

And please add `putenv('PIMCORE_SKIP_DOTENV_FILE=true');` for skip loading .env file

jdreesen replied:
> Isn't above line is enough for setting app env?

Probably depends on how it's read ;)

yariksheptykin replied:
Yeah exactly. Normally `putenv` should be enough. BUT php has a wierd handling of ENV vars. Some are set by webserver (or fast-cgi) in $_SERVER, some in ENV / getenv via shell for cli. People expect the envs be set at different places. To make sure everybody gets these vars, I set them everywhere. But putenv should be enough in most cases, so if you find it more appropriate we can use this.

dvesh3 commented on config/packages/test/config.yaml:
```diff
@@ -1,18 +1,2 @@
 imports:
     - { resource: ../dev/config.yaml }
-
-doctrine:
```
why removed it? we should not use dev db for running tests

yariksheptykin replied:
Good question @dvesh3 ! I forgot why I had to fo this back then. I agree, we should keep the test and dev databases separately. To keep this PR small I shortcut test dns to dev db. I can update compose.yml with a test db setup in a follow up, if you are interested, or desribe in readme how one could do it.

aryaantony92 replied:
> Thanks for looking at this PR @aryaantony92 ! I worked your feedback and suggestions in. Please take another look at my changes and the discussion regarding composer commands.

@yariksheptykin Thanks. Everything looks good. But, can you please update the test db setup also in this PR itself ? Because, we updated the steps in read me to run codeception

yariksheptykin committed:
Use test db for executing tests

yariksheptykin committed:
Use union type

yariksheptykin committed:
bootstrap PIMCORE_SKIP_DOTENV_FILE with true

jdreesen commented on docker-compose.yaml:
```diff
@@ -31,6 +31,12 @@ services:
+            #
+            # Warning:
+            # We strongly recommend changing this environment variable to use a separate database for tests.
+            # Eexecuting tests against the default database may lead to unexpected results.
```
```suggestion
            # Executing tests against the default database may lead to unexpected results.
```

yariksheptykin committed:
Jacob is the best

Co-authored-by: Jacob Dreesen <jacob@hdreesen.de>

aryaantony92 commented on tests/unit.suite.yml:
```diff
@@ -0,0 +1 @@
+namespace: \\App\\Tests\\Unit
```
```suggestion
suite_namespace: App\Tests\Unit

```

aryaantony92 commented on tests/functional.suite.yml:
```diff
@@ -0,0 +1,2 @@
+namespace: \\App\\Tests\\Functional
```
```suggestion
suite_namespace: App\Tests\Functional
```

aryaantony92 commented on codeception.yml:
```diff
@@ -0,0 +1,12 @@
+    tests: tests
+    output: tests/_output
+    support: tests/_support
+    data: tests
```
```suggestion
    data: tests/Support/Data
```

aryaantony92 replied:
```suggestion
    data: tests/Support/Data
```

aryaantony92 commented on codeception.yml:
```diff
@@ -0,0 +1,12 @@
+namespace: \\App\\Tests
```
```suggestion
namespace: App\Tests
```

aryaantony92 replied:
```suggestion
namespace: \App\Tests
```

aryaantony92 replied:
```suggestion
namespace: App\Tests
```

aryaantony92 reviewed (commented):
@yariksheptykin 
Thanks for the PR. The tests look good. But, please change the namespaces as suggested below to follow the codeception 5 standard.

Also capitalize the suite names from 'unit.suite.yaml' to 'Unit.suite.yml' and 'functional.suite.yml' to 'Functional.suite.yml'. And change the folder names also from '_support' to 'Support', 'functional' to 'Functional' and 'unit' to 'Unit'

aryaantony92 commented on composer.json:
```diff
@@ -25,6 +25,14 @@
+  "autoload-dev": {
+    "psr-4": {
+      "App\\Tests\\": "tests/",
+      "App\\Tests\\Unit\\": "tests/unit",
```
The entries for Unit, Functional and Support can be removed as we follow the correct namespaces

aryaantony92 commented on .github/workflows/pimcore-skeleton.yml:
```diff
@@ -92,3 +92,6 @@ jobs:
         php vendor/bin/pimcore-install -n
+
+        # Execute codeception tests.
+        docker compose exec -T php composer test
```
@yariksheptykin May I know why did you call the test from composer ? We can use the command 'docker compose exec php vendor/bin/codecept run -vv' here to run the tests.

yariksheptykin replied:
@aryaantony92 sure. I use composer script here as an alias for running the tests, just to make the command invocation shorter. I personally find it easier to memorize `composer test`. To my experience in projects there is a bunch of "commands" that one needs to run to get stuff done. These commands are often spread across readme.md, some shell-scripts / Makefiles, and some undocumented conventions. I prefer to define all these commands in composer.json as scripts. It is then easier to memorize & operate the project dev setup.

Because I also prefer to have CI test as much as possible I used this shortcut in the CI too.

This is however a personal pref and has nothing to do with setting up codeception. If you prefer to use codecept binary directly, I can update the code.

aryaantony92 replied:
Thank you for the info. But, could you please use the codeception command itself as we follow this pattern in core and other bundles.

yariksheptykin replied:
Sure, I will update the PR.

aryaantony92 commented on codeception.yml:
```diff
@@ -0,0 +1,12 @@
+paths:
+    tests: tests
+    output: tests/_output
+    support: tests/_support
```
```suggestion
    support: tests/Support
```

yariksheptykin committed:
Update names for test directories

yariksheptykin commented:
Thanks for looking at this PR @aryaantony92 ! I worked your feedback and suggestions in. Please take another look at my changes and the discussion regarding composer commands.

yariksheptykin committed:
Inline composer test command

aryaantony92 commented:
> Thanks for looking at this PR @aryaantony92 ! I worked your feedback and suggestions in. Please take another look at my changes and the discussion regarding composer commands.

@yariksheptykin Thanks. Everything looks good. But, can you please update the test db setup also in this PR itself ? Because, we updated the steps in read me to run codeception

aryaantony92 commented:
> > Thanks for looking at this PR @aryaantony92 ! I worked your feedback and suggestions in. Please take another look at my changes and the discussion regarding composer commands.
> 
> @yariksheptykin Thanks. Everything looks good. But, can you please update the test db setup also in this PR itself ? Because, we updated the steps in read me to run codeception

Just a gentle reminder

yariksheptykin committed:
Compose a test setup

yariksheptykin committed:
Mount over var dirs in test

yariksheptykin committed:
Block php on db health

yariksheptykin committed:
Fix ci

yariksheptykin commented:
> > Thanks for looking at this PR @aryaantony92 ! I worked your feedback and suggestions in. Please take another look at my changes and the discussion regarding composer commands.
> 
> @yariksheptykin Thanks. Everything looks good. But, can you please update the test db setup also in this PR itself ? Because, we updated the steps in read me to run codeception

yariksheptykin commented:
> @yariksheptykin Thanks. Everything looks good. But, can you please update the test db setup also in this PR itself ? Because, we updated the steps in read me to run codeception

@aryaantony92  I updated the PR to include a test database setup in compose.

jdreesen commented on docker-compose.yaml:
```diff
@@ -6,14 +6,23 @@ services:
+            # The service is considered healthy when:
+            # - connection to pimcore db can be established, and
+            # - we can run a simple SQL query.
+            test: [ "CMD-SHELL", "mysql -h db -u$$MYSQL_USER -p$$MYSQL_PASSWORD $$MYSQL_DATABASE -e 'SHOW TABLES;'" ]
```
What about using this included script for a health check? https://github.com/MariaDB/mariadb-docker/blob/master/10.7/healthcheck.sh

yariksheptykin replied:
Interesting. I was not aware of this healthcheck-script, thanks for sharing.

The main purpose of the healthcheck on db container is to block the startup of other containers until the database is set up. The database is set up when the container entrypoint created pimcore schema & user and the server has restarted to accept connections from "outside" (not only at localhost). After this happens we should be able to install pimcore without a need for "wait-for-it" helpers.

I am not familiar with the healthcheck.sh baked into the mariadb container. Can we achieve the same goal for us?
I also was not able to find a version if this script for [mysql](https://github.com/docker-library/mysql). Does it also have one? Otherwise this healthcheck will only work out of the box for mariadb images.

in any case, I would like to move this discussion to a follow up MR, where we can focus on setting up db properly. To me this discussion departs from the original intent of this MR, which was to help developers get stated with writing tests in their projects. Let's wrap this bit first, and then focus on setting up the database.

yariksheptykin commented:
@aryaantony92 A gentle reminder that I updated setting up the database as you requested. Please take a another look at the PR.

aryaantony92 commented on docker-compose.yaml:
```diff
@@ -31,19 +40,56 @@ services:
             PHP_IDE_CONFIG: serverName=localhost
+            # Feed installer configuration via ENV variables.
+            # See: https://pimcore.com/docs/pimcore/current/Development_Documentation/Getting_Started/Advanced_Installation_Topics.html#page_Advanced-Installation-Topics
+            PIMCORE_INSTALL_ADMIN_USERNAME: pimcore
```
Thank you for the changes. Could you please move 'PIMCORE_INSTALL_ADMIN_USERNAME: pimcore' from 'php:environment' to test-php;environment so that the user can choose admin username freely while installing pimcore. https://github.com/pimcore/skeleton/blob/88a9ab4f9cf220a22fb484247421fe04c975f213/README.md?plain=1#L41. Rest looks good

yariksheptykin committed:
Prompt for admin username and password at install

aryaantony92 commented:
@yariksheptykin Thank you!

yariksheptykin commented:
@aryaantony92 @jdreesen @dvesh3  it was a pleasure collaborating with you. Thanks for your constructive feedback.
